In [1]:
import pandas as pd
import re

In [2]:
airbnb_df = pd.read_csv("/Users/laracacador/Documents/Ironhack/Week_3/Project/airnb.csv", encoding='utf8')

In [3]:
airbnb_df.head()

,Title,Detail,Date,Price(in dollar),Offer price(in dollar),Review and rating,Number of bed
0,"Chalet in Skykomish, Washington, US",Sky Haus - A-Frame Cabin,Jun 11 - 16,306.00,229.00,4.85 (531),4 beds
1,"Cabin in Hancock, New York, US",The Catskill A-Frame - Mid-Century Modern Cabin,Jun 6 - 11,485.00,170.00,4.77 (146),4 beds
2,"Cabin in West Farmington, Ohio, US",The Triangle: A-Frame Cabin for your city retreat,Jul 9 - 14,119.00,522.00,4.91 (515),4 beds
3,"Home in Blue Ridge, Georgia, US",*Summer Sizzle* 5 Min to Blue Ridge* Pets* Hot...,Jun 11 - 16,192.00,348.00,4.94 (88),5 beds
4,"Treehouse in Grandview, Texas, US",Luxury Treehouse Couples Getaway w/ Peaceful V...,Jun 4 - 9,232.00,196.00,4.99 (222),1 queen bed


In [4]:
airbnb_df.columns.tolist()

['Title',
 'Detail',
 'Date',
 'Price(in dollar)',
 'Offer price(in dollar)',
 'Review and rating',
 'Number of bed']

## Visualizing the data before cleaning: 

In [5]:
airbnb_df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 953 entries, 0 to 952
Data columns (total 7 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Title                   953 non-null    object
 1   Detail                  953 non-null    object
 2   Date                    953 non-null    object
 3   Price(in dollar)        953 non-null    object
 4   Offer price(in dollar)  166 non-null    object
 5   Review and rating       947 non-null    object
 6   Number of bed           953 non-null    object
dtypes: object(7)
memory usage: 52.2+ KB


## Checking columns: Title, Price(in dollar) and Offer price(in dollar)

In [6]:
airbnb_df[["Title", "Price(in dollar)", "Offer price(in dollar)"]].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 953 entries, 0 to 952
Data columns (total 3 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Title                   953 non-null    object
 1   Price(in dollar)        953 non-null    object
 2   Offer price(in dollar)  166 non-null    object
dtypes: object(3)
memory usage: 22.5+ KB


##based on these findings, from 953 entries:
-> columns "title" and "price(in dollar)" have zero null values, however
-> offer price have only 166 non-null entries which correspond to only ~17,4% of the column being populated ((166*100) / 953) (or 82.6% missing)
-> title, price(in dollar) and offer price(in dollar) = str (object) type


## Checking Duplicates

In [7]:
duplicates_total = airbnb_df.duplicated().sum()
print(f"Total of duplicated rows is: {duplicates_total}")
duplicate_rows = airbnb_df[airbnb_df.duplicated()]
duplicate_rows.head(15)

Total of duplicated rows is: 34


,Title,Detail,Date,Price(in dollar),Offer price(in dollar),Review and rating,Number of bed
189,"Hotel in Tepoztlán, Mexico",Posada Sarita,Jun 3 - 10,50.00,NaN,4.56 (176),3 beds
327,"Home in Baião, Portugal",CasasdoCasinhotoD-Rustic House (pool; river view),Jun 12 - 17,131.00,NaN,4.81 (43),1 double bed
379,"Guesthouse in Perranwell Station, UK",The Folly - Cornwall,Jun 18 - 23,111.00,NaN,5.0 (11),1 bed
392,"Home in Audierne, France",Villa Trouz Ar Mor,Jul 1 - 6,135.00,NaN,4.85 (13),2 beds
419,"Home in Bad Sankt Leonhard im Lavanttal, Austria",1A chalet resting crickets + hiking with indoo...,Jun 22 - 27,123.00,NaN,4.67 (110),3 beds
420,"Vacation home in Punta Prima, Spain",Beachfront apartment with perfect seaview,Jun 1 - 6,231.00,NaN,4.85 (204),4 beds
470,"Earthen home in Mueang Chiang Mai District, Th...",Leafy Greens Chiangmai : Mushroom M2,Jun 13 - 20,22.00,NaN,4.87 (210),1 bunk bed
525,"Hotel in Tambon Ratsada, Thailand",Hotel Room with Private Plunge Pool Siray Bay#2,Jun 13 - 18,71.00,NaN,5.0 (13),1 bed
560,"Home in Kecamatan Selat, Indonesia",The Nude House,Jul 10 - 15,305.00,NaN,4.68 (19),2 beds
561,"Villa in Kathu, Thailand",Tiny Poolvilla in the heart of Phuket,Jul 25 - 30,119.00,NaN,5.0 (98),1 king bed


##total of 34 duplicates

## Cleaning duplicates

In [8]:
airbnb_df.drop_duplicates(inplace=True)

In [9]:
duplicates_total = airbnb_df.duplicated().sum()
print(f"Total of duplicated rows is: {duplicates_total}")
duplicate_rows = airbnb_df[airbnb_df.duplicated()]
duplicate_rows.head(15)

Total of duplicated rows is: 0


,Title,Detail,Date,Price(in dollar),Offer price(in dollar),Review and rating,Number of bed


In [10]:
airbnb_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 919 entries, 0 to 952
Data columns (total 7 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Title                   919 non-null    object
 1   Detail                  919 non-null    object
 2   Date                    919 non-null    object
 3   Price(in dollar)        919 non-null    object
 4   Offer price(in dollar)  166 non-null    object
 5   Review and rating       913 non-null    object
 6   Number of bed           919 non-null    object
dtypes: object(7)
memory usage: 57.4+ KB


In [11]:
airbnb_df.shape

(919, 7)

##we can observe that the duplicated rows were removed, hence the number 919 (953 - 34). we can also infer that offer price non-null count stayed at 166 before and after,because none of the removed duplicated rows had a value in offer price

In [12]:
duplicate_rows["Offer price(in dollar)"].notna().sum()

np.int64(0)

##we can now confirm that: none of the removed duplicate rows in the column 'offer price' had a value

## Splitting the column 'review and rating' in 2 different columns 

In [13]:
airbnb_df.columns = [column.lower().replace(" ", "_").replace("(", "").replace(")", "") for column in airbnb_df.columns]
airbnb_df

,title,detail,date,pricein_dollar,offer_pricein_dollar,review_and_rating,number_of_bed
0,"Chalet in Skykomish, Washington, US",Sky Haus - A-Frame Cabin,Jun 11 - 16,306.00,229.00,4.85 (531),4 beds
1,"Cabin in Hancock, New York, US",The Catskill A-Frame - Mid-Century Modern Cabin,Jun 6 - 11,485.00,170.00,4.77 (146),4 beds
2,"Cabin in West Farmington, Ohio, US",The Triangle: A-Frame Cabin for your city retreat,Jul 9 - 14,119.00,522.00,4.91 (515),4 beds
3,"Home in Blue Ridge, Georgia, US",*Summer Sizzle* 5 Min to Blue Ridge* Pets* Hot...,Jun 11 - 16,192.00,348.00,4.94 (88),5 beds
4,"Treehouse in Grandview, Texas, US",Luxury Treehouse Couples Getaway w/ Peaceful V...,Jun 4 - 9,232.00,196.00,4.99 (222),1 queen bed
...,...,...,...,...,...,...,...
948,Guest suite in Lake Worth,The Garden Suite at the Historic Chateau Bleu,May 1 - 6,141.00,NaN,4.85 (82),1 king bed
949,Room in Kissimmee,Private Room in Orlando/Kissimmee (2Twins) T1,May 1 - 6,43.00,NaN,4.9 (300),1 king bed
950,Guesthouse in San Antonio,Casa Alexander,May 1 - 6,73.00,NaN,4.92 (282),5 beds
951,Apartment in Englewood,BEACH FRONT Unit w/ Pool Sandy Toe's # 5,May 7 - 12,233.00,NaN,4.96 (318),4 beds


In [14]:
airbnb_df_copy = airbnb_df.copy()
airbnb_df_copy[['review_and_rating', 'amount_of_answers']] = airbnb_df_copy.review_and_rating.str.split(" ", expand=True)

In [15]:
airbnb_df_copy

,title,detail,date,pricein_dollar,offer_pricein_dollar,review_and_rating,number_of_bed,amount_of_answers
0,"Chalet in Skykomish, Washington, US",Sky Haus - A-Frame Cabin,Jun 11 - 16,306.00,229.00,4.85,4 beds,(531)
1,"Cabin in Hancock, New York, US",The Catskill A-Frame - Mid-Century Modern Cabin,Jun 6 - 11,485.00,170.00,4.77,4 beds,(146)
2,"Cabin in West Farmington, Ohio, US",The Triangle: A-Frame Cabin for your city retreat,Jul 9 - 14,119.00,522.00,4.91,4 beds,(515)
3,"Home in Blue Ridge, Georgia, US",*Summer Sizzle* 5 Min to Blue Ridge* Pets* Hot...,Jun 11 - 16,192.00,348.00,4.94,5 beds,(88)
4,"Treehouse in Grandview, Texas, US",Luxury Treehouse Couples Getaway w/ Peaceful V...,Jun 4 - 9,232.00,196.00,4.99,1 queen bed,(222)
...,...,...,...,...,...,...,...,...
948,Guest suite in Lake Worth,The Garden Suite at the Historic Chateau Bleu,May 1 - 6,141.00,NaN,4.85,1 king bed,(82)
949,Room in Kissimmee,Private Room in Orlando/Kissimmee (2Twins) T1,May 1 - 6,43.00,NaN,4.9,1 king bed,(300)
950,Guesthouse in San Antonio,Casa Alexander,May 1 - 6,73.00,NaN,4.92,5 beds,(282)
951,Apartment in Englewood,BEACH FRONT Unit w/ Pool Sandy Toe's # 5,May 7 - 12,233.00,NaN,4.96,4 beds,(318)


In [16]:
airbnb_df_copy["amount_of_answers"] = airbnb_df_copy["amount_of_answers"].str.replace(r'[()]', '', regex=True)

In [17]:
airbnb_df_copy

,title,detail,date,pricein_dollar,offer_pricein_dollar,review_and_rating,number_of_bed,amount_of_answers
0,"Chalet in Skykomish, Washington, US",Sky Haus - A-Frame Cabin,Jun 11 - 16,306.00,229.00,4.85,4 beds,531
1,"Cabin in Hancock, New York, US",The Catskill A-Frame - Mid-Century Modern Cabin,Jun 6 - 11,485.00,170.00,4.77,4 beds,146
2,"Cabin in West Farmington, Ohio, US",The Triangle: A-Frame Cabin for your city retreat,Jul 9 - 14,119.00,522.00,4.91,4 beds,515
3,"Home in Blue Ridge, Georgia, US",*Summer Sizzle* 5 Min to Blue Ridge* Pets* Hot...,Jun 11 - 16,192.00,348.00,4.94,5 beds,88
4,"Treehouse in Grandview, Texas, US",Luxury Treehouse Couples Getaway w/ Peaceful V...,Jun 4 - 9,232.00,196.00,4.99,1 queen bed,222
...,...,...,...,...,...,...,...,...
948,Guest suite in Lake Worth,The Garden Suite at the Historic Chateau Bleu,May 1 - 6,141.00,NaN,4.85,1 king bed,82
949,Room in Kissimmee,Private Room in Orlando/Kissimmee (2Twins) T1,May 1 - 6,43.00,NaN,4.9,1 king bed,300
950,Guesthouse in San Antonio,Casa Alexander,May 1 - 6,73.00,NaN,4.92,5 beds,282
951,Apartment in Englewood,BEACH FRONT Unit w/ Pool Sandy Toe's # 5,May 7 - 12,233.00,NaN,4.96,4 beds,318


## Missing values for the pricein_dollar and title columns

In [18]:
airbnb_df[['pricein_dollar', 'title']].isna().sum()

pricein_dollar    0
title             0
dtype: int64

##confirms that there is no missing values for pricein_dollar and title columns

In [19]:
## Creating a shot before cleaning title and price

In [20]:
airbnb_df_original = airbnb_df.copy()  

## Cleaning pricein_dollar and title columns

### formatting title

In [22]:
# "type in location", for example '(type:)Apartment in (location:)Tulum, Mexico'
# n=1 because i want just to separate from the first "in" and not in another possible others 'in' alongside
airbnb_df[["property_type", "location"]] = airbnb_df["title"].str.split(" in ", n=1, expand=True) 
airbnb_df.head()

,title,detail,date,pricein_dollar,offer_pricein_dollar,review_and_rating,number_of_bed,property_type,location
0,"Chalet in Skykomish, Washington, US",Sky Haus - A-Frame Cabin,Jun 11 - 16,306.00,229.00,4.85 (531),4 beds,Chalet,"Skykomish, Washington, US"
1,"Cabin in Hancock, New York, US",The Catskill A-Frame - Mid-Century Modern Cabin,Jun 6 - 11,485.00,170.00,4.77 (146),4 beds,Cabin,"Hancock, New York, US"
2,"Cabin in West Farmington, Ohio, US",The Triangle: A-Frame Cabin for your city retreat,Jul 9 - 14,119.00,522.00,4.91 (515),4 beds,Cabin,"West Farmington, Ohio, US"
3,"Home in Blue Ridge, Georgia, US",*Summer Sizzle* 5 Min to Blue Ridge* Pets* Hot...,Jun 11 - 16,192.00,348.00,4.94 (88),5 beds,Home,"Blue Ridge, Georgia, US"
4,"Treehouse in Grandview, Texas, US",Luxury Treehouse Couples Getaway w/ Peaceful V...,Jun 4 - 9,232.00,196.00,4.99 (222),1 queen bed,Treehouse,"Grandview, Texas, US"


#### in the dataset, 'location' can be only city, or city+country, or city+state+country
#### we need to know how many ' , ' exists to confirm which formats they are

In [23]:
airbnb_df["location"].str.count(",").value_counts()

location
1    554
0    240
2    125
Name: count, dtype: int64

#554 'city, country'

#240 with just 'city'

#125 'city, state, country'

In [26]:
location_parts = airbnb_df["location"].str.split(",", expand=True)
location_parts.head()

,0,1,2
0,Skykomish,Washington,US
1,Hancock,New York,US
2,West Farmington,Ohio,US
3,Blue Ridge,Georgia,US
4,Grandview,Texas,US


#splitting the 'location' column where there is a ' , ' , to have later separate city state and country columns based on position

In [27]:
location_parts[2].isna().sum()

np.int64(794)

#checking how many rows have a value in index [2] after the split

#expected: 794 nulls (919 - 125), since only 125 rows had 2 commas

#city + state + country format

In [32]:
# to remove any space 

location_parts = location_parts.apply(lambda col: col.str.strip())

In [33]:
# city is always the first split value (index 0) 
# + every row has at least a city name
# + confirmed earlier that column 0 has no missing values

airbnb_df["city"] = location_parts[0]

In [34]:
# country = column 2 when it exists (city+state+country format); 
# otherwise -> falls back to column 1 (city+country format, no state)

airbnb_df["country"] = location_parts[2].where(location_parts[2].notna(), location_parts[1])

In [35]:
# state only exists when there was a 3rd part (column 2 present)
# in the city+country rows, column 1 is already used as country 
# so state is correctly left as NaN there
airbnb_df["state"] = location_parts[1].where(location_parts[2].notna())

In [36]:
location_parts[location_parts[2].notna()].head()

,0,1,2
0,Skykomish,Washington,US
1,Hancock,New York,US
2,West Farmington,Ohio,US
3,Blue Ridge,Georgia,US
4,Grandview,Texas,US


In [37]:
airbnb_df[["city", "state", "country"]].isna().sum()

city         0
state      794
country    240
dtype: int64

#### recap:
#### city: 0 nulls = every row has a city
#### state: 794 nulls (919 - 125, only 3-part rows have a state)
#### country: 240 nulls = rows that only had a city, with no country in 
#### the original data = matches the earlier count of 0-comma rows

In [38]:
airbnb_df[["city", "state", "country"]].dtypes

city       object
state      object
country    object
dtype: object

### formatting price

In [40]:
airbnb_df.columns.tolist()

['title',
 'detail',
 'date',
 'pricein_dollar',
 'offer_pricein_dollar',
 'review_and_rating',
 'number_of_bed',
 'property_type',
 'location',
 'city',
 'country',
 'state']

In [41]:
airbnb_df['pricein_dollar'].dtype

dtype('O')

In [45]:
float("1,463.00")  #errrrrrrrrrror, commas are enemy of float, a dot is bff

ValueError: could not convert string to float: '1,463.00'

In [42]:
airbnb_df["price"] = airbnb_df["pricein_dollar"].str.replace(",", "", regex=False).astype(float)

In [43]:
airbnb_df["price"].describe()

count     919.000000
mean      171.696409
std       142.093744
min        16.000000
25%        85.000000
50%       135.000000
75%       220.500000
max      1463.000000
Name: price, dtype: float64

#### since the original column has data type=str, the goal was to:
#### 1. remove the ',' of the thousands separator (ex: 1,000)
#### 2. regex=False bc we want pandas to treat "," as a literal character to find and replace
#### 3. astype(float) to convert it to a number, after having remove the comma bc otherwise we would get a ValueError

# Establishing relation between price and property_type

#### Average price by property_type

In [50]:
# reliable vs low-volume property types
price_by_type = (airbnb_df.groupby("property_type")["price"].agg(["count", "mean", "median"]).round(2).sort_values("mean", ascending=False))
price_by_type

,count,mean,median
property_type,,,
Island,1,469.00,469.0
Villa,59,287.59,209.0
Home,114,241.74,185.0
Treehouse,37,235.49,207.0
Dome,11,235.27,291.0
Cottage,34,218.15,147.5
Boat,4,202.25,203.0
Vacation home,4,196.50,193.0
Windmill,1,196.00,196.0


In [57]:
# property types with count >= 10 , so we can trust the listings in order to trust the average
reliable_types = price_by_type[price_by_type["count"] >= 10]
reliable_types

,count,mean,median
property_type,,,
Villa,59,287.59,209.0
Home,114,241.74,185.0
Treehouse,37,235.49,207.0
Dome,11,235.27,291.0
Cottage,34,218.15,147.5
Cabin,115,193.50,171.0
Condo,88,168.18,144.5
Guest suite,13,163.85,138.0
Bungalow,11,137.55,74.0


In [54]:
# property types with count < 10
# just one or two listings = these shouldn't be read as reliable patterns
low_volume_types = price_by_type[price_by_type["count"] < 10]
low_volume_types

,count,mean,median
property_type,,,
Island,1,469.00,469.0
Boat,4,202.25,203.0
Vacation home,4,196.50,193.0
Windmill,1,196.00,196.0
Chalet,9,190.33,105.0
Barn,5,184.20,163.0
Townhouse,4,181.75,145.0
Tower,1,166.00,166.0
Nature lodge,5,160.20,159.0


#### insights:

#### - Villa has the highest average price (~287$),
#### however Home have just a smaller average price (~242$) the double amount of listings (114 > 59). In that sense, the amount of money made by the type Home seems to be greater than the one made by Villa type, bc it represents the largest share of the market by listing count;

#### - Higher volume =/= higher price, bc:
####   -- Cabin accounts for 115 listings (the biggest in the dataset) but with an average of ~193$,
####   -- Apartment with 115 listings only with an average of 125$,
####   -- Condo with 88 listings and average of 168$ 

#### - Mean vs median: cottage, and bungalow have greater disparity between mean and median, which means that the outliers in the list are pulling the average up, so this is worth taking in account for possible price adjustments(?)

In [61]:
price_by_type["estimated_total"] = price_by_type["mean"] * price_by_type["count"]
price_by_type.sort_values("estimated_total", ascending=False)

,count,mean,median,estimated_total
property_type,,,,
Home,114,241.74,185.0,27558.36
Cabin,115,193.50,171.0,22252.50
Villa,59,287.59,209.0,16967.81
Condo,88,168.18,144.5,14799.84
Apartment,115,124.68,93.0,14338.20
Treehouse,37,235.49,207.0,8713.13
Room,81,101.81,92.0,8246.61
Cottage,34,218.15,147.5,7417.10
Tiny home,41,136.27,123.0,5587.07


In [59]:
price_by_type.sort_values("estimated_total", ascending=False).head(10)

,count,mean,median,estimated_total
property_type,,,,
Home,114,241.74,185.0,27558.36
Cabin,115,193.50,171.0,22252.50
Villa,59,287.59,209.0,16967.81
Condo,88,168.18,144.5,14799.84
Apartment,115,124.68,93.0,14338.20
Treehouse,37,235.49,207.0,8713.13
Room,81,101.81,92.0,8246.61
Cottage,34,218.15,147.5,7417.10
Tiny home,41,136.27,123.0,5587.07


#### Home bringing most value to the market